apps/reaper/lib/reaper/data_extract/extract_step.ex

```diff
@@ -0,0 +1,82 @@
+defmodule Reaper.DataExtract.ExtractStep do
+  @moduledoc """
+  This module processes extract steps as defined in a dataset definition.  After
+  iterating through the steps, accumulating any destination values in the assigns block
+  it is assumed the final step will be http (at this time) which returns a data stream
+  """
+  require Logger
+  alias Reaper.DataSlurper
+  alias Reaper.UrlBuilder
+  alias Reaper.Decoder
+
+  def execute_extract_steps(dataset, steps) do
+    Enum.reduce(steps, %{}, fn step, acc ->
+      execute_extract_step(dataset, step, acc)
+    end)
+  end
+
+  defp execute_extract_step(dataset, step, assigns_accumulator) do
+    step = Map.put(step, :assigns, Map.merge(step.assigns, assigns_accumulator))
+    process_extract_step(dataset, step)
+  rescue
+    error ->
+      Logger.error(Exception.format(:error, error, __STACKTRACE__))
+      reraise "Unable to process #{step.type} step for dataset #{dataset.id}.", __STACKTRACE__
+  end
+
+  defp process_extract_step(dataset, %{type: "http"} = step) do
+    {body, headers} = evaluate_body_and_headers(step)
+
+    UrlBuilder.decode_http_extract_step(step)
+    |> DataSlurper.slurp(dataset.id, headers, step.context.protocol, step.context.action, body)
+    |> Decoder.decode(dataset)
+    |> Stream.with_index()
```

Comment:
I've been thinking about this part - are we painting ourselves into a corner by making it so the "http" step always has to be last? Do we have cases we might need in the future, where we go do an arbitrary http endpoint to assign something? I don't think so, but it's worth considering.

Reply:
Talked through this.